snuba/clusters/cluster.py

```diff
@@ -395,11 +395,50 @@ def _get_storage_set_cluster_map() -> Dict[StorageSetKey, ClickhouseCluster]:
     return _STORAGE_SET_CLUSTER_MAP
 
 
+def _build_partitioned_cluster(cluster: Mapping[str, Any]) -> ClickhouseCluster:
+    return ClickhouseCluster(
+        host=cluster["host"],
+        port=cluster["port"],
+        user=cluster.get("user", "default"),
+        password=cluster.get("password", ""),
+        database=cluster.get("database", "default"),
+        http_port=cluster["http_port"],
+        storage_sets={storage_tuple[0] for storage_tuple in cluster["storage_sets"]},
+        single_node=cluster["single_node"],
+        cluster_name=cluster["cluster_name"] if "cluster_name" in cluster else None,
+        distributed_cluster_name=cluster["distributed_cluster_name"]
+        if "distributed_cluster_name" in cluster
+        else None,
+        cache_partition_id=cluster.get("cache_partition_id"),
+        query_settings_prefix=cluster.get("query_settings_prefix"),
+    )
+
+
+_PARTITIONED_STORAGE_SET_CLUSTER_MAP: Dict[
+    Tuple[StorageSetKey, int], ClickhouseCluster
+] = {}
+
+
+def _get_partitioned_storage_set_cluster_map() -> Dict[
+    Tuple[StorageSetKey, int], ClickhouseCluster
+]:
+    if len(_PARTITIONED_STORAGE_SET_CLUSTER_MAP) == 0:
+        for cluster in settings.PARTITIONED_CLUSTERS:
+            for storage_set_tuple in cluster["storage_sets"]:
+                _PARTITIONED_STORAGE_SET_CLUSTER_MAP[
+                    (StorageSetKey(storage_set_tuple[0]), storage_set_tuple[1])
+                ] = _build_partitioned_cluster(cluster)
+
+    return _PARTITIONED_STORAGE_SET_CLUSTER_MAP
+
+
 class UndefinedClickhouseCluster(SerializableException):
     pass
 
 
-def get_cluster(storage_set_key: StorageSetKey) -> ClickhouseCluster:
+def get_cluster(
+    storage_set_key: StorageSetKey, partition_id: Optional[int] = None
```

Comment:
How is this function supposed to be used? Please add a docstring as it's imported in quite a few places. Is None is supposed to be passed only for unpartitioned, and an int for partitioned? Since they are handled completely different you may want to consider splitting it into 2 functions instead.

Another idea I'll just throw out there would be to make partition_id required, and treat everything as partitioned. An unpartitioned storage set is nothing more than a storage set with 1 partition as far as most of Snuba is concerned.

Reply:
We'll probably also need a `get_clusters(storage_set_key)` to retrieve them all in this module. There's a bunch of pieces of code (e.g. migrations, admin, etc) that will likely need to fetch all the clusters of a storage set and iterate through them.

Reply:
I like the idea of making partition id required. I'll add it and see what the others also think

Reply:
> Another idea I'll just throw out there would be to make partition_id required, and treat everything as partitioned. An unpartitioned storage set is nothing more than a storage set with 1 partition as far as most of Snuba is concerned. Signature of this function would then have `partition_id: int = 0`

I've thought about this and don't like the idea. Based on the way I think we should handle topologies, you'd need to map every physical partition to a storage (https://www.notion.so/sentry/Generic-Metrics-Slicing-9c206e02f92a4abca5948d2e21c6c00d#a1c689d0674c4671ba646ad98d1cf44d).

If you don't like that idea, then we can discuss in that doc or tomorrow's design meeting, but I'd prefer to leave the unpartitioned clusters alone for the time being

Reply:
Oh no, just to clarify I am not suggesting to change PARTITIONED_CLUSTERS or CLUSTERS anything about that interface. Only to hide the difference between PARTITIONED_CLUSTERS and CLUSTERS inside this function and make them the same to the outside world. Basically everything from CLUSTERS gets treated as partition 0 in here.

Reply:
Hm, I'm not sure I quite follow. I think if we have it so that partition id is not optional (i.e. for the default case we pass in 0 always), the function logic would have to involve deciding whether to look into CLUSTERS or PARTITIONED_CLUSTERS. Because the function doesn't know if the storage set key we passed in is one that is not partitioned (resides in CLUSTERS) or partitioned, with multiple partition ids (resides in PARTITIONED_CLUSTERS). I think that would interfere with the clear distinction between CLUSTERS and PARTITIONED_CLUSTERS, which is defined in a comment now in settings.

Reply:
CLUSTERS vs PARTITIONED_CLUSTERS only exists for one reason - to keep settings backwards compatible and CLUSTERS unchanged for those who don't need to care about extra slices. There is no real reason for all the different parts of Snuba to keep this distinction, in fact it adds complexity everywhere.

In this current implementation our whole code base has to know about the difference between partitioned and unpartitioned clusters. What I'm proposing is we have this function decide which one to look in. The benefit of that is that no one else has to. Right now every caller of this module needs to know the difference between partitioned and unpartitioned clusters. Why not move that job into this module and make them exactly the same to all the other parts of Snuba.

Reply:
so @lynnagara is your thought that we do something like

```
_SEEN_PARTITIONED_STORAGE_SET_KEYS = extract_storage_set_keys(_PARTITIONED_STORAGE_SET_CLUSTER_MAP)

def get_cluster(storage_set_key, partition_id = 0):
   if storage_set_key in _SEEN_PARTITIONED_STORAGE_SET_KEYS:
     get from partitioned_clusters
   else:
     assert partition_id == 0
     get from unpartitioned_clusters
```

I'm fine with that if you feel strongly but I aesthetically/subjectively prefer Riya's current approach -- I don't see how it pollutes the codebase more than yours given both approaches rely on the optionality of the argument in both to avoid adding `partition_id=n` everywhere.

Reply:
OK, I'd like to continue as-is, but I agree that we are invalidating the documentation (currently) on lines 444-448 of this file and it should be corrected to clearly explain the new behavior.

Reply:
Is this a blocker or can we follow up on this separately. If it is not a blocker, can we open a follow up task and get the current implementation merged? I would like to use the functionality being introduced in this PR for https://github.com/getsentry/snuba/pull/3172

Reply:
Right, I'll update the function docstring (apologies for not getting that done earlier)

Reply:
Not a blocker, but I think it would be a lot better if this function took a logical partition instead of a physical partition and did the mapping in one place. I don't think there's a reason we need to call `map_logical_partition_to_physical_partition()` everywhere that then calls this function instead of just doing it just once here.

Reply:
I see, is this referring to how this code will integrate with #3172? Do you think that change (logical to physical partition mapping inside get_cluster()) should be done before this PR gets merged?

Reply:
No, go ahead and merge, we can figure out those details later.

Reply:
I think I understand the idea about how every call to get_cluster() now would need to "think" about whether or not it is dealing with a partitioned or unpartitioned cluster (and whether or not to pass a slice id or not). Should changing this be a follow up task? Or should it be done before merging?